Set default MemorySwappiness when adapt

hqhq opened the pull request:
It makes the inspect result consistent between cli and REST api
when MemorySwappiness is not set.

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

runcom commented:
Tests failures seems related

hqhq committed:
Move security opt adaption to adapt function

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@runcom Updated.

estesp commented:
ping @vieux : does this fix the issue you mentioned recently re: memoryswappiness default?

cpuguy83 commented:
This is not the first time we've gone through such a change on this field.

hqhq commented:
@cpuguy83 What change exactly are you mentioning? I know this field used to be `int64` and was changed to `*int64`, but this PR is more like a fix instead of a change. WDYT?

cpuguy83 commented:
Changing the defaults for MemorySwapiness -- just pointing out this has been a problem multiple times.

runcom commented:
I think the default if not provided will default to -1 as it was previously. This patch just move the initialization to -1 back in the code path right @hqhq?

estesp commented:
More correctly, it handles non-Docker clients (e.g. API users like swarm/compose) from causing "unset" to translate to 0, rather than "unset" -> -1, which is what should happen. I think most of the confusion was around catching all the places that instantiation was performed and making sure default != 0, given 0 is a setting, not a default in this case.

estesp commented:
sorry @hqhq ..this needs a rebase now

hqhq commented:
@estesp was right, this PR will fix problem like https://github.com/docker/swarm/issues/1411
The problem we have before was when user used REST API, the default value for `swappiness` is not set to container's hostConfig, only set to `container.command` and passed to libcontainer, which will cause problem like the swarm issue.

@runcom @estesp Rebased.

hqhq committed:
Set default MemorySwappiness when adapt

It makes the inspect result consistent between cli and REST api
when MemorySwappiness is not set.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

jimmyxian commented:
@hqhq Thanks for doing this.
I think we need set `swappiness` default value -1 in docker daemon side.

thaJeztah commented:
@hqhq @jimmyxian was this issue introduced in docker 1.9 in Swarm, or already present in docker 1.8.3? Asking so that we can decide if this should be considered for a 1.9.2 release (if we decide to release that)

hqhq commented:
@thaJeztah It's exist since Docker v1.5

runcom commented:
I think my PR #18261 will be also affected by having adaptContainerSettings always called? Is this correct @hqhq ?

hqhq commented:
@runcom Yeah, you can see I changed your test cases. You expected API without hostConfig (or just without these fields) has `nil` for these pointer fields, while CLI without these options has default value (not `nil`). Are there any specific reasons that I missed? I think it's better to make them consistent. And the reason we want API got the same default value is the same as we want CLI got the default.

runcom commented:
@hqhq perfect

estesp commented:
ugh..definitely lots of rebases being required the last 24 hours..sorry @hqhq !

estesp commented:
LGTM

cpuguy83 commented:
LGTM

thaJeztah commented:
@estesp @cpuguy83 sorry, I lost track of all the "default value" changes; is this an user-facing change in the API? If so, can you add the right impact labels?

estesp commented:
I don't believe there is any need for API or user-facing impact--this is really a bug fix for a hole in the setup of containers via certain paths which weren't defaulting properly to the "unset" value (a pointer to -1 in this case), meaning they were getting "default to zero" as a setting for swappiness, which is an actual swappiness value, which of course is a bug unless the user asked for swappiness = 0.

thaJeztah commented:
Thanks for checking @estesp!
